File: src/liffier/detector.py
"""Response analysis to detect successful path traversal."""

from __future__ import annotations

import re
from dataclasses import dataclass

# Signatures that indicate a successful file read per common target files
SIGNATURES: dict[str, list[str]] = {
    "/etc/passwd": ["root:x:0:0:", "root:*:0:0:", "daemon:", "nobody:"],
    "/etc/shadow": ["root:$", "root:!", "root:*:"],
    "/etc/hosts": ["127.0.0.1", "localhost"],
    "/etc/hostname": [],  # any non-error content
    "/proc/self/environ": ["PATH=", "HOME=", "USER="],
    "/proc/version": ["Linux version"],
    "/proc/self/cmdline": [],
    "win.ini": ["[fonts]", "[extensions]", "[files]"],
    "boot.ini": ["[boot loader]", "[operating systems]", "WINDOWS"],
    "C:\\Windows\\System32\\drivers\\etc\\hosts": ["127.0.0.1", "localhost"],
    "web.xml": ["<web-app", "<servlet"],
    ".env": ["APP_KEY=", "DB_PASSWORD=", "DB_HOST="],
}

# Generic error patterns that indicate the traversal did NOT work
ERROR_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"(404|not\s*found)", re.IGNORECASE),
    re.compile(r"(403|forbidden|access\s*denied)", re.IGNORECASE),
    re.compile(r"(400|bad\s*request)", re.IGNORECASE),
    re.compile(r"(500|internal\s*server\s*error)", re.IGNORECASE),
    re.compile(r"<html[^>]*>.*<title>.*error.*</title>", re.IGNORECASE | re.DOTALL),
]


@dataclass
class Detection:
    """Result of analyzing a response for traversal success."""

    hit: bool
    confidence: str  # "high", "medium", "low"
    reason: str
    matched_signature: str = ""


def analyze_response(
    status_code: int,
    body: str,
    target_file: str,
    baseline_length: int | None = None,
) -> Detection:
    """Analyze an HTTP response to determine if traversal succeeded.

    Args:
        status_code: HTTP response status code.
        body: Response body text.
        target_file: The file being targeted (e.g. /etc/passwd).
        baseline_length: Length of a known-bad response for comparison.
    """
    # Non-200 is almost certainly not a hit
    if status_code >= 400:
        return Detection(hit=False, confidence="high", reason=f"HTTP {status_code}")

    # Check for known error patterns in body
    for pat in ERROR_PATTERNS:
        if pat.search(body[:2000]):
            return Detection(hit=False, confidence="medium", reason=f"Error pattern: {pat.pattern[:40]}")

    # Check for file-specific signatures
    normalized = target_file.replace("\\", "/").strip("/").split("/")[-1]
    for sig_file, sigs in SIGNATURES.items():
        sig_normalized = sig_file.replace("\\", "/").strip("/").split("/")[-1]
        if normalized == sig_normalized or target_file.endswith(sig_normalized):
            for sig in sigs:
                if sig in body:
                    return Detection(
                        hit=True,
                        confidence="high",
                        reason=f"Signature match for {sig_file}",
                        matched_signature=sig,
                    )
            # File matched but no signature found
            if sigs:  # has signatures but none matched
                break
            # No specific signatures (like /etc/hostname) - check body isn't empty
            if len(body.strip()) > 0 and status_code == 200:
                return Detection(
                    hit=True,
                    confidence="medium",
                    reason=f"Non-empty 200 response for {sig_file} (no specific signature)",
                )

    # Baseline length comparison - significant difference suggests different content
    if baseline_length is not None and len(body) != baseline_length:
        diff_ratio = abs(len(body) - baseline_length) / max(baseline_length, 1)
        if diff_ratio > 0.3 and len(body) > 50:
            return Detection(
                hit=True,
                confidence="low",
                reason=f"Response length differs from baseline ({len(body)} vs {baseline_length}, {diff_ratio:.0%} diff)",
            )

    # 200 with content but no signature match
    if status_code == 200 and len(body.strip()) > 100:
        return Detection(
            hit=False,
            confidence="low",
            reason="200 OK but no signature match - review manually",
        )

    return Detection(hit=False, confidence="medium", reason="No indicators of success")

File: internal/detector/detector.go
package detector

import (
	"math"
	"strings"
)

// Detection is the result of analyzing a response for traversal success.
type Detection struct {
	Hit              bool
	Confidence       string // "high", "medium", "low"
	Reason           string
	MatchedSignature string
}

// fileSignatures maps target file basenames to content signatures.
var fileSignatures = map[string][]string{
	"passwd":  {"root:x:0:0:", "root:*:0:0:", "daemon:", "nobody:"},
	"shadow":  {"root:$", "root:!", "root:*:"},
	"hosts":   {"127.0.0.1", "localhost"},
	"environ": {"PATH=", "HOME=", "USER="},
	"version": {"Linux version"},
	"win.ini": {"[fonts]", "[extensions]", "[files]"},
	"boot.ini": {"[boot loader]", "[operating systems]"},
	".env":    {"APP_KEY=", "DB_PASSWORD=", "DB_HOST="},
	"web.xml": {"<web-app", "<servlet"},
}

// errorPatterns in response bodies that indicate failure.
var errorPatterns = []string{
	"404", "not found", "403", "forbidden", "access denied",
	"400", "bad request", "500", "internal server error",
}

// Analyze checks an HTTP response for signs of successful path traversal.
func Analyze(statusCode int, body string, targetFile string, baselineLength int) Detection {
	if statusCode >= 400 {
		return Detection{Hit: false, Confidence: "high", Reason: "HTTP " + strings.TrimSpace(string(rune(statusCode+'0')))}
	}

	// Extract basename from target file
	basename := targetFile
	if idx := strings.LastIndex(targetFile, "/"); idx >= 0 {
		basename = targetFile[idx+1:]
	}
	if idx := strings.LastIndex(basename, "\\"); idx >= 0 {
		basename = basename[idx+1:]
	}

	// Check for error patterns in body (first 2000 chars)
	snippet := body
	if len(snippet) > 2000 {
		snippet = snippet[:2000]
	}
	lower := strings.ToLower(snippet)
	for _, pat := range errorPatterns {
		if strings.Contains(lower, pat) {
			return Detection{Hit: false, Confidence: "medium", Reason: "Error pattern: " + pat}
		}
	}

	// Check file-specific signatures
	for file, sigs := range fileSignatures {
		if basename == file || strings.HasSuffix(targetFile, file) {
			for _, sig := range sigs {
				if strings.Contains(body, sig) {
					return Detection{
						Hit:              true,
						Confidence:       "high",
						Reason:           "Signature match for " + file,
						MatchedSignature: sig,
					}
				}
			}
			// File matched but no signatures found
			if len(sigs) > 0 {
				break
			}
			if len(strings.TrimSpace(body)) > 0 && statusCode == 200 {
				return Detection{
					Hit:        true,
					Confidence: "medium",
					Reason:     "Non-empty 200 for " + file + " (no specific signature)",
				}
			}
		}
	}

	// Baseline length comparison
	if baselineLength > 0 && len(body) != baselineLength {
		diff := math.Abs(float64(len(body) - baselineLength))
		ratio := diff / math.Max(float64(baselineLength), 1)
		if ratio > 0.3 && len(body) > 50 {
			return Detection{
				Hit:        true,
				Confidence: "low",
				Reason:     "Response length anomaly vs baseline",
			}
		}
	}

	return Detection{Hit: false, Confidence: "medium", Reason: "No indicators"}
}
